addons/l10n_it_edi/tests/test_ir_mail_server.py

```diff
@@ -0,0 +1,115 @@
+            return self.server._create_invoice_from_mail(content, filename, 'fake@address.be')
+
+    def _get_test_content(self, filename):
+        """ Get the content of a test file inside this module """
+        path = get_module_resource('l10n_it_edi', 'tests', 'expected_xmls', filename)
+        with tools.file_open(path, mode='rb') as test_file:
+            return test_file.read()
+
+    def test_attachment_invoice_already_received(self):
+        """ Test that the second time we are receiving a PEC mail with the same attachment, the second is discarded """
+        content = self.test_content.encode('utf-8')
+        for result in [True, False]:
+            self.assertEqual(result, bool(self._create_invoice(content, 'AZ123123123_12345.xml')))
+
+    def test_receive_vendor_bill(self):
+        """ Test a sample e-invoice file from https://www.fatturapa.gov.it/export/documenti/fatturapa/v1.2/IT01234567890_FPR01.xml """
+        # Simulate the reception of an einvoice
+        invoices = self._create_invoice(self.invoice_content, self.invoice_filename)
+        self.assertTrue(bool(invoices))
+
+    def _test_receipt(self, receipt_type, source_state, destination_state):
+        """ Test a receipt from the ones in the module's test files """
+
+        # Simulate the 'sent' state of the move, even if we didn't actually send an email in this test
+        move = self.env['account.move'].search([('ref', '=', self.invoice_ref)])
+        move.l10n_it_send_state = source_state
+
+        # Create the attachment and the account.edi.document just as if we sent the email
+        attachment = self.env['ir.attachment'].search([('name', '=', self.invoice_filename)])
+        edi_document = self.env['account.edi.document'].create({
+            'edi_format_id': self.env.ref('l10n_it_edi.edi_fatturaPA').id,
+            'move_id': move.id,
+            'state': 'sent',
+            'attachment_id': attachment.id
+        })
+
+        # Create a fake receipt from the test file
+        receipt_filename = 'IT01234567890_FPR01_%s_001.xml' % receipt_type
+        receipt_content = self._get_test_content(receipt_filename).decode('utf-8')
+
+        create_mail_attachment = namedtuple('Attachment', ('fname', 'content', 'info'))
+        receipt_mail_attachment = create_mail_attachment(receipt_filename, receipt_content, {})
+
+        # Simulate the arraival of the receipt
+        with patch.object(self.server._cr, 'commit', return_value=None):
+            self.server._message_receipt_invoice(receipt_type, receipt_mail_attachment)
+
+        # Check the Destination state of the edi_document
+        self.assertTrue(destination_state, edi_document.state)
+
+    def test_ricevuta_consegna(self):
+        """ Test a receipt adapted from https://www.fatturapa.gov.it/export/documenti/messaggi/v1.0/IT01234567890_11111_RC_001.xml """
+
+        # Prepare the invoice in the system
+        self.test_receive_vendor_bill()
```

Comment:
It is weird to call a test method inside of another test method. Didn't you want to use a set-up instead?

Reply:
This is the reason, the tests are:
1) (test_receive_vendor_bill) Receive one vendor bill XML from mail -> OK
2) (test_attachment_invoice_already_received) Receive two times the same vendor bill XML from email -> OK, ERROR
3) (test_ricevuta_consegna) Receive a OK receipt XML from mail for a previously imported vendor bill -> OK
4) (test_decorrenza_termini) Receive a "N-OK" receipt XML from mail for a previously imported vendor bill -> OK

If I make a setup with the first vendor bill already in the database, then I have no way to do test 1) (apart from making two test classes or whatever).
I'll change the name test_attachment_invoice_already_received -> test_receive_same_vendor_bill_twice for clarity.

Reply:
The issue is that `test_ricevuta_consegna` and `test_decorrenza_termini` have a dependency on `test_receive_vendor_bill`. If the later fails the tests, it is impossible and useless to test the other two.

If you add this in the `setUpClass`
```python
        cls.invoices = self._create_invoice(self.invoice_content, self.invoice_filename)
```

Then
```python
    def test_receive_vendor_bill(self):
        """ Test a sample e-invoice file from https://www.fatturapa.gov.it/export/documenti/fatturapa/v1.2/IT01234567890_FPR01.xml """
        self.assertTrue(bool(self.invoices))

    def test_ricevuta_consegna(self):
        """ Test a receipt adapted from https://www.fatturapa.gov.it/export/documenti/messaggi/v1.0/IT01234567890_11111_RC_001.xml """
        self._test_receipt('RC', 'sent', 'delivered')

    def test_decorrenza_termini(self):
        """ Test a receipt adapted from https://www.fatturapa.gov.it/export/documenti/messaggi/v1.0/IT01234567890_11111_DT_001.xml """
        self._test_receipt('DT', 'delivered', 'delivered_expired')
```

Reply:
Your solution wouldn't solve the problem: if self._create_invoice(...) throws, the setUpClass fails, the test class is teared down, and all the tests are skipped.

So I decided to set up more test data in the setUpClass() method by hand just for the receipt, and avoid calling the test_receive_vendor_bill() in the tests. I hope it's better this way.